homeassistant/components/light/__init__.py

```diff
@@ -225,6 +252,41 @@ async def async_handle_light_on_service(light, call):
 
             preprocess_turn_on_alternatives(hass, params)
 
+        supported_color_modes = light.supported_color_modes
+        # Backwards compatibility: if an RGBWW color is specified, convert to RGB + W
+        # for legacy lights
+        if ATTR_RGBW_COLOR in params:
+            legacy_supported_color_modes = light._light_internal_supported_color_modes
+            if (
+                COLOR_MODE_RGBW in legacy_supported_color_modes
+                and not supported_color_modes
+            ):
+                rgbw_color = params.pop(ATTR_RGBW_COLOR)
+                params[ATTR_RGB_COLOR] = rgbw_color[0:2]
+                params[ATTR_WHITE_VALUE] = rgbw_color[3]
+
+        # If a color is specified, convert to the color space supported by the light
+        # Backwards compatibility: Fall back to hs color if light.supported_color_modes
+        # is not implemented
+        if ATTR_HS_COLOR or ATTR_RGB_COLOR or ATTR_XY_COLOR in params:
```

Comment:
This tests if `ATTR_HS_COLOR` is truthy, `ATTR_RGB_COLOR` is truthy, or `ATTR_XY_COLOR in params`. The first two are defined so this code is always executed?
```suggestion
        if ATTR_HS_COLOR in params or ATTR_RGB_COLOR in params or ATTR_XY_COLOR in params:
```

Reply:
> I would expect Google, Alexa and Homekit requiring updates because of this PR

Does it mean those integrations will need to be updated to handle the new colour mode, rgbw and rgbww?

Reply:
Oops, tests are needed for the new code

Reply:
Couldn't you use intersection here?

Reply:
Not sure if more readable:

```python
if params.keys() & {ATTR_HS_COLOR, ATTR_RGB_COLOR, ATTR_XY_COLOR}:
```

You can also do

```python
if any(key in params for key in (ATTR_HS_COLOR, ATTR_RGB_COLOR, ATTR_XY_COLOR)):
```